src/attribution/onINP.ts

```diff
@@ -25,38 +26,109 @@ import {
+            longestLoAFsList.splice(i, 0, entry);
+            break;
+          }
+        }
+        if (longestLoAFsList.length > MAX_LOAFS_TO_CONSIDER) {
+          longestLoAFsList.splice(MAX_LOAFS_TO_CONSIDER);
+        }
+      }
+    }
+  });
+};
+
 const attributeINP = (metric: INPMetric): void => {
-  if (metric.entries.length) {
-    const longestEntry = metric.entries.sort((a, b) => {
-      // Sort by: 1) duration (DESC), then 2) processing time (DESC)
-      return (
-        b.duration - a.duration ||
-        b.processingEnd -
-          b.processingStart -
-          (a.processingEnd - a.processingStart)
-      );
-    })[0];
-
-    // Currently Chrome can return a null target for certain event types
-    // (especially pointer events). As the event target should be the same
-    // for all events in the same interaction, we pick the first non-null one.
-    // TODO: remove when 1367329 is resolved
-    // https://bugs.chromium.org/p/chromium/issues/detail?id=1367329
-    const firstEntryWithTarget = metric.entries.find((entry) => entry.target);
-
-    (metric as INPMetricWithAttribution).attribution = {
-      eventTarget: getSelector(
-        firstEntryWithTarget && firstEntryWithTarget.target,
-      ),
-      eventType: longestEntry.name,
-      eventTime: longestEntry.startTime,
-      eventEntry: longestEntry,
-      loadState: getLoadState(longestEntry.startTime),
-    };
-    return;
+  const sortedEntries = metric.entries.sort((a, b) => {
+    return a.processingStart - b.processingStart;
+  });
+
+  const firstEntry = sortedEntries[0];
+  const lastEntry = sortedEntries[sortedEntries.length - 1];
+  const renderTime = firstEntry.startTime + firstEntry.duration;
+
+  let longAnimationFrameEntry;
+  for (const loaf of longestLoAFsList) {
+    const loafEnd = loaf.startTime + loaf.duration;
+    if (
+      firstEntry.startTime === loaf.firstUIEventTimestamp ||
+      (loafEnd <= renderTime && loafEnd >= firstEntry.processingStart)
+    ) {
+      longAnimationFrameEntry = loaf;
+      break;
+    }
   }
-  // Set an empty object if no other attribution has been set.
-  (metric as INPMetricWithAttribution).attribution = {};
+
+  // If the browser supports the Long Animation Frame API and a
+  // `long-animation-frame` entry was found matching this interaction,
+  // use that entry's `renderTime` since it could be more accurate (and
+  // it accounts for non-event listener work such as timers or other
+  // pending tasks that could get run before rendering starts). If not,
+  // use the `processingEnd` value of the last entry in the list, which
+  // should match the `renderTime` value in most cases.
+  const renderStart = longAnimationFrameEntry
+    ? longAnimationFrameEntry.renderStart
+    : lastEntry.processingEnd;
+
+  // The first entry may not have a target defined, so use the first
+  // one found in the entry list.
+  const firstEntryWithTarget = metric.entries.find((entry) => entry.target);
```

Comment:
Repeating comment: I would sort/split metric.entries on those which have interactionID first, so as to ignore early events such as pointerover.

Reply:
Good call. Done.